src/resources/uart_esp32.cc

```diff
@@ -17,95 +17,584 @@
+      abort();
+    }
   }
-};
 
-MODULE_IMPLEMENTATION(uart, MODULE_UART);
+  uart()->enable_interrupt_index(UART_TOIT_INTR_TXFIFO_EMPTY);
+}
+
+UART_ISR_INLINE uint8_t TxBuffer::read_break() {
+  IsrSpinLocker locker(&spinlock_);
+
+  uint8_t break_length = transfer_header_.break_length_;
+  transfer_header_.break_length_ = 0; // To avoid it being read again and the break continuing...
+  return break_length;
+}
+
+UART_ISR_INLINE uint8_t* TxBuffer::read(size_t *received, size_t max_length) {
+  IsrSpinLocker locker(&spinlock_);
+
+  if (transfer_header_.remaining_data_length_ == 0 && transfer_header_.break_length_ == 0) {
+    // No more data in the latest package. Need to read header
+    size_t header_received;
+    size_t read = 0;
+    do { // As the header is multiple bytes, it could be split over the edge of the ring buffer
+      void* header_data = xRingbufferReceiveUpToFromISR(ring_buffer(), &header_received, sizeof(TxTransferHeader) - read);
+      if (!header_data) {
+        // TxBuffer is empty
+        return null;
+      }
+      memcpy(reinterpret_cast<uint8_t*>(&transfer_header_) + read, header_data, header_received);
+      // No blocking calls ever on TxBuffer, so ignore last parameter
+      vRingbufferReturnItemFromISR(ring_buffer(), header_data, null);
+      read += header_received;
+    } while (read < sizeof(TxTransferHeader));
+  }
+
+  size_t maximum_to_receive = max_length;
+  if (transfer_header_.remaining_data_length_ < maximum_to_receive) maximum_to_receive = transfer_header_.remaining_data_length_;
+
+  if (maximum_to_receive == 0) {
+    return null;
+  }
+
+  auto buffer = static_cast<uint8_t*>(xRingbufferReceiveUpToFromISR(ring_buffer(), received, maximum_to_receive));
+  transfer_header_.remaining_data_length_ -= *received;
+  return buffer;
+}
+
+size_t TxBuffer::free_size() {
+  size_t ringbuffer_free = xRingbufferGetCurFreeSize(ring_buffer());
+  if (ringbuffer_free < sizeof(TxTransferHeader)) return 0;
+  return ringbuffer_free - sizeof(TxTransferHeader);
+}
+
+void UART_ISR_INLINE RxBuffer::send(uint8_t* buffer, uint32_t length) {
+  // We are never blocking on the ring buffer
+  xRingbufferSendFromISR(ring_buffer(), buffer, length, null);
+}
+
+void RxBuffer::read(uint8_t* buffer, uint32_t length) {
+  size_t received;
+  size_t read = 0;
+  do {
+    auto from_ring_buffer = static_cast<uint8_t*>(xRingbufferReceiveUpTo(ring_buffer(), &received, 0, length - read));
+    memcpy(buffer + read, from_ring_buffer, received);
+    return_buffer(from_ring_buffer);
+    read += received;
+  } while (read < length);
+  uart()->enable_read_interrupts();
+}
+
+bool UartResource::receive_event(word* data) {
+  return xQueueReceive(queue(), data, 0);
+}
+
+UartResource::~UartResource() {
+  disable_interrupt_index(UART_TOIT_ALL_INTR_MASK);
+  if (interrupt_handle_) esp_intr_free(interrupt_handle_);
+
+  vQueueDelete(queue());
+
+  // Empty tx buffer.
+  uint32_t len;
+  do {
```

Comment:
Not sure this is good.  It will hang for an indefinite time, no?

Reply:
As this is interacting directly with the uart controller it will actually complete quickly. It depends on the baud rate. The max size is 128 bytes, so for 9600 baud it empties in a little over a second, worst case.
The reason that this is here, is that if you for example use jaguar and deploys a new version while a transmit is in progress, it will not work at all. I could not find a better way to do it, due to hardware limitations.